fblevel.8
.\" Copyright (c) 2000 Stephan Leemburg
.\"
.\"  This program is free software; you can redistribute it and/or
.\"  modify it under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version
.\"  2 of the License, or (at your option) any later version.
.\"
.\" $Log: fblevel.8,v $
.\" Revision 1.1.1.1  2001/12/07 11:31:50  sleemburg
.\" Initial CVS import of the unreleased pmud-0.8 to apmud (new project name
.\" because of a name clash at sourceforge.net).
.\"
.\" Revision 1.1  2000/05/11 14:52:40  stephan
.\" Initial revision
.\"
.\"
.Dd March 28, 2000
.Dt FBLEVEL 8
.Os "LinuxPPC pmud"
.Sh NAME
.Nm fblevel
.Nd sets the display brightness level on a notebook
.Sh SYNOPSIS
.Nm fblevel
.Op on
.Op off
.Op <level>
.Sh DESCRIPTION
.Nm fblevel
is a program, with which you can set the Framebuffer Backlight level on notebooks. 
The minimum level is 0 and the maximum is 31. 
<level>
adjustment will only set the backlight level, whereas 
<on> or
<off>
will power on or off parts of the display controller as well.
.Pp 
With no option, 
.Nm fblevel
will report the current backlight level. 
.Sh BUGS AND CHANGE REQUESTS
Please email your bug reports or change requests to \fB<pmud-bugs@jvc.nl>\fP. 
.Sh FILES
.nf
/dev/fb
.fi
.Sh AUTHOR
Stephan Leemburg <stephan@jvc.nl>.
.Sh SEE ALSO
.Xr pmud(8)


debian/changelog
pmac-utils (1.1.3-27+ports) unreleased; urgency=medium

  * Upload to unreleased for powerpc and ppc64.
  * Bump compat level to 9 in debian/compat.
  * Drop #define __USE_GNU from nwnvsetenv.c.
  * Fix syntax error in autoboot.sgml.

 -- John Paul Adrian Glaubitz <glaubitz@physik.fu-berlin.de>  Tue, 08 Jan 2019 17:48:12 +0300

pmac-utils (1.1.3-27) unstable; urgency=medium

  * Incremental packaging updates for Standards-Version: 3.9.7:
    - Record our source format (1.0) in debian/source/format.
    - Stop redundantly setting DH_COMPAT in debian/rules.
    - Add build-indep and build-arch targets to debian/rules.
    - Use dh_prep instead of dh_clean -k in debian/rules.
    - Add comment-only watch file to document upstream state.
    - Use ${misc:Depends} in debian/control for binaries.
    - Remake debian/copyright with a DEP-5 compatible format.

 -- Adam Conrad <adconrad@0c3.net>  Tue, 05 Apr 2016 12:48:28 -0600

pmac-utils (1.1.3-26) unstable; urgency=medium

  * Add myselt to Uploaders with Michael's consent.
  * Rename to pmac-utils, powerpc-utils is another project.
  * Stop building on little-endian arches (LP: #1398163)
  * Drop ancient and obsolete Conflicts/Replaces.

 -- Adam Conrad <adconrad@0c3.net>  Sun, 27 Mar 2016 13:42:07 -0600

powerpc-utils (1.1.3-25) unstable; urgency=medium

  * Update maintainer address. Closes: #762790.
  * Fix compiler warnings.

 -- Michael Schmitz <schmitzmic@gmail.com>  Wed, 29 Oct 2014 20:28:32 +1100

powerpc-utils (1.1.3-24.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Add ppc64el to the Architecture: field. Closes: #738140.

 -- Aurelien Jarno <aurel32@debian.org>  Sat, 20 Sep 2014 22:09:45 +0200

powerpc-utils (1.1.3-24) unstable; urgency=low

  * Fix fixed postinstall handling of adjtime setup (copy /etc/adjtime if it
    exists, create new file with the correct data otherwise). 
    Closes: #528443.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 13 May 2009 20:04:24 +1100

powerpc-utils (1.1.3-23) unstable; urgency=low

  * Fix broken -f option for fnset. Check for presence of ADB devices and bail 
    out on USB based machines.
    Closes: #474814.

  * Fix postinstall handling of adjtime setup (copy /etc/adjtime if it
    exists, create new file otherwise). 
    Closes: #528443.

  * Update bootsched manpage to clarify that it should only be used on 
    Oldworld PowerMacs.
    Closes: #407295.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 13 May 2009 19:34:24 +1100

powerpc-utils (1.1.3-22) unstable; urgency=low

  * Fix problem with nvram reads segfaulting on efika, reported by Robert
    Millan <rm@aybabtu.com> and patched by same.
    Closes: #431116.

  * Honor DEB_BUILD_OPTS=noopt, patch by Robert Millan <rm@aybabtu.com>.
    Closes: #431117.

  * Add Aurelien Gerome <ag@roxor.cx> as co-maintainer.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed,  4 Jul 2007 17:21:40 +0100

powerpc-utils (1.1.3-21) unstable; urgency=low

  * Fix problem with truncated nvram reads/writes, reported by Adam
    Lackorzynski <adam@os.inf.tu-dresden.de>.
    Closes: #413131.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed,  7 Mar 2007 15:57:34 +0100

powerpc-utils (1.1.3-20) unstable; urgency=low

  * Change section to 'utils' to match override file.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Tue, 30 Jun 2006 15:39:34 +0100

powerpc-utils (1.1.3-19) unstable; urgency=low

  * The 'upload lost' release :-(

  * Incorporate newworld autoboot patch by Rich Johnson.

  * Fix description formatting (thanks to Nick Boullis <nboullis@debian.org>
    for the alert).
    Closes: #336972.

  * Add ppc64 arch (just in case biarch powerpc isn't enough).
    Closes: #360797.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Tue, 30 Jun 2006 15:39:34 +0100

powerpc-utils (1.1.3-18) unstable; urgency=low

  * add -D_GNU_SOURCE to Makefile in order to help with 64 bit pointer
    size on ppc64.
    Closes: #326394.

  * Fix problem introduced in -12 (/etc/adjtime converted to symlink 
    into /var fs): Convert symlink to plain copy now.
    Closes: #332926.

  * Fix char signedness bug in nvsetvol (patch by Eugen Dedu
    <Eugen.Dedu@pu-pm.univ-fcomte.fr>.
    Closes: #334547

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 19 Oct 2005 18:07:34 +0200

powerpc-utils (1.1.3-17) unstable; urgency=low

  * Fix for iBook G4 failure to save nvram data (as manifest by failure to
    properly store the boot chime volume). Patch by Joerg Dorchain
    <joerg@dorchain.net>.
    Closes: #321655, #320566.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 08 Aug 2005 14:40:34 +0200

powerpc-utils (1.1.3-16) unstable; urgency=low

  * Brown Paper Bag release: Added bootsched util (for CUDA based Powermacs)
    the source for which I had buried for over three years now.

  * Finally, actually install the autoboot binary (reported missing by
    Herve Eychenne <rv@eychenne.org>).
    Closes: #317021.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 06 Jul 2005 09:42:44 +0200

powerpc-utils (1.1.3-15) unstable; urgency=low

  * Merge back patches from ubuntu (thanks to Adam Conrad <adconrad@0c3.net>
    for the hint): fix linuxdoc-tools b-d, drop adb_mouse.h header. Update 
    description to include autoboot and lsprop tools.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 09 May 2005 12:31:24 +0100

powerpc-utils (1.1.3-14) unstable; urgency=low

  * Fixed nvsetvol failure (infinite loop reading NVRAM) on OldWorld
    machines (reported by Christophe Donatsch <cdonatsch@tele2.ch>).
    Closes: #304493.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Sun, 01 May 2005 11:58:37 +0100

powerpc-utils (1.1.3-13) unstable; urgency=low

  * Fixed nvsetenv's devfsd bug (reported by Roger Leigh <rleigh@debian.org>).
    Closes: #300946.

  * Added autoboot util (for CUDA based Powermacs)


fblevel.c
/*
 * ----------------------------------------------------------------------------
 * fblevel - sets the fb device brightness level on Apple Powerbooks
 *
 * Copyright 2000 Stephan Leemburg <stephan@jvc.nl>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 * ----------------------------------------------------------------------------
 * $Log: fblevel.c,v $
 * Revision 1.1.1.1  2001/12/07 11:31:50  sleemburg
 * Initial CVS import of the unreleased pmud-0.8 to apmud (new project name
 * because of a name clash at sourceforge.net).
 *
 * Revision 1.3  2001/11/09 10:55:29  stephan
 * use "pmu.h"
 *
 * Revision 1.2  2000/10/09 14:16:33  stephan
 * use /dev/fb0 in stead of /dev/fb
 *
 * Revision 1.1  2000/05/11 14:48:46  stephan
 * Initial revision
 *
 * ----------------------------------------------------------------------------
 */

#include <stdio.h>
#include <ctype.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

#include <linux/fb.h>
#include <linux/pmu.h>

#ifndef FBIOBLANK
#define FBIOBLANK      0x4611          /* 0 or vesa-level+1 */
#endif

static void fbon(int on);
static int set_fblevel(int level);
static int get_fblevel(void);
static void usage(char *program);

int main(int argc, char **argv)
{
	char *program = strrchr(*argv, '/');

	program = program ? program + 1 : *argv;

	switch(argc)
	{
	case 1:
		printf("%d\n", get_fblevel());
		break;
	case 2:
		if(!strcasecmp(*++argv, "on")
		|| !strcasecmp(*argv, "off"))
			fbon(strcasecmp(*argv, "off"));
		else
		{
			register char *p;

			for(p = *argv; *p && isdigit(*p); p++)
				;
			if(*p)
			{
				usage(program);
				return 1;
			}
			if(!set_fblevel(atoi(*argv)))
				return 1;
		}
		break;
	default:
		usage(program);
		return 1;

	}
	return 0;
}

static void fbon(int on)
{
	static int fd = -1;
	if(fd<0)
	{
		fd = open("/dev/fb0", O_RDONLY);
		if(fd<0)
		{
			perror("/dev/fb0");
			return;
		}
	}
	(void)ioctl(fd, FBIOBLANK, !on);
}

static int set_fblevel(int level)
{
	int fd, ret=0;
	if ( (fd = open("/dev/pmu", O_RDONLY)) < 0 ) {
		perror("/dev/pmu");
		goto out;
	}
	if(ioctl(fd, PMU_IOC_SET_BACKLIGHT, &level) == -1) {
		perror("ioctl PMU_IOC_SET_BACKLIGHT");
		goto out;
	}
	ret = 1;	/* success */
out:
	close(fd);
	return ret;
}

static int get_fblevel(void)
{
	int fd, level=0;
	if ( (fd = open("/dev/pmu", O_RDONLY)) < 0 ) {
		perror("/dev/pmu");
		goto out;
	}
	if(ioctl(fd, PMU_IOC_GET_BACKLIGHT, &level) == -1) {
		perror("ioctl PMU_IOC_GET_BACKLIGHT");
		goto out;
	}
out:
	close(fd);
	return level;
}

static void usage(char *program)
{
	fprintf(stderr, 
		"usage: %s [on|off|<level>]\n"
		"       on     : powers display on\n"
		"       off    : powers display off\n"
		"       <level>: set display brightness to <level>\n"
		"                (level is an integer)\n"
		"no argument prints the current brightness level\n",
                program
		);
}
